### PR TITLE
Add page for finding and merging duplicate lots

### LINKED ISSUE #5166: Feature Request: detect duplicate lots
Building off of https://github.com/IMA-WorldHealth/bhima/issues/5165, we probably want a way to detect duplicate lots to merge them.  This would be based on the lot label being the same for a product of the same inventory item.

### PR DESCRIPTION
**Pull request by jmcameron**
Create a page where operators can find and merge potential duplicate lots. 

**Note** that this PR adds a new page under the Stock navigation menu:  "Find Duplicate Lots".  In order to use or test this, be sure to edit the user role permissions and enable the new page (under the Stock section).

Closes #5166

**Testing**
- Use a production database such as IMCK
- Start Bhima
- In Adminstration > Role Management , enable user access to the "Find Duplicate Lots" page.
- Go to Stock > Find Duplicate Lots
   - Note the number of duplicates found (see the grid footer)
   - Click on the Action menu of any row and click on  "Find duplicate lots".  (I'm not 100% happy about the wording here; I'm open to suggestions).
   - A modal will pop up.  Note that form will show all lots that have matching lot labels AND inventory UUIDs.  Only one lot is shown for all the lots that match it.
   - Select the lot to keep and the lot(s) to merge
   - Click on the [Merge] button
   - When you return to the main grid, notice the offending row is now gone (since there is only one of them)

### CODE REVIEW

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.html]: FYI - to make this grid full height, use the class `grid-full-height`.  This accounts for the "filter-bar".

**Review by jniles — COMMENTED** [review]
@jmcameron since you asked for a review of the functionality, I won't dive into the code yet.

Overall, looks really good for a first cut.  Some suggestions:

  1. Instead of "Find Lots" in the action menu, let's just call it "Merge Lots".  To me, this is intuitive: instead of finding the lots, we've found them, now we are opening the modal to merge them.
  2. Can you put a loading indicator on the loading button when merging lots?  It seems like we should be waiting for that to be done on the server, then closing the modal, then reloading the page.  It dismisses almost immediately, which surprised me.  Most other modals that do something in the system (trial balance, credit note, etc) hang out until the action is performed with a loading indicator.
 
Here is what that looks like on a "good 2G network" that I emulated.
MOYdGh0e9j

Again, haven't taken a look at the code, but it surprised me how quick that was!

**Comment by jmcameron** [comment]
@jniles This is now ready for a real review.   

**QUESTION**:  Should we remove the "Find Duplicate Lots" in the Lots registry action menu?

**Comment by jniles** [comment]
> QUESTION: Should we remove the "Find Duplicate Lots" in the Lots registry action menu?

No, I do not think so.  Let's wait until a couple of releases, then we'll remove it.  I think that folks are much more inclined to absent-mindedly look at that registry (which they are using all the time) than they will be for this new module.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.html]: A meta comment - I find it a little strange this file path is `lots-duplicates/duplicate-lots.html`.  Is a reason why the folder name and the file name are different?

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jniles) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.html]: Since we don't have any filters on this registry, you should include this change.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.js]: Let's just remove this.

File: client/src/modules/stock/lots-duplicates/templates/action.cell.html
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/templates/action.cell.html]: You don't need the `{{rowRenderIndex}}`.  This is a testing anti-pattern (click in row by it's position in the grid).  Since there isn't an end to end test for this yet, you can just remove this binding.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.js]: You'll want to set "hasError" to true here.
```js
.catch((err) => {
  hasError = true;
  Notify.handleError(err);
})
```

This tells the grid to show an error message.  `Notify.handleError` doesn't do that part automatically - it just makes the notification bar come down.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.js]: You can probably remove all references to `gridApi`.  I don' think you need it or use it anywhere.  I can't see a scenario where you would need it either.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.js]: These comments should probably be updated to reflect the true nature of the controller.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jniles) [inline client/src/modules/stock/lots-duplicates/duplicate-lots.html]: Do you need the `ui-grid-grouping` here?

**Review by jniles — APPROVED** [review]
Functionally, this looks good to me.  I've made a few clean up comments, but none of them are serious.  Once addressed, this PR can be merged.

**Comment by jniles** [comment]
bors delegate+

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.html]: Done

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.js:21]: I wanted this to sort next to the 'lots' file folder uner 'stock', hence the name.  I plan to rename the files to 'lots-duplicates.*' to match the directory name.   'duplicate-lots' would be more readable, but would make it a bit harder to find with the file explorer.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.html]: I wanted this to sort next to the 'lots' file folder uner 'stock', hence the name. I plan to rename the files to 'lots-duplicates.*' to match the directory name. 'duplicate-lots' would be more readable, but would make it a bit harder to find with the file explorer.

Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.html]: Files renamed.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.js]: Removed

File: client/src/modules/stock/lots-duplicates/templates/action.cell.html
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/templates/action.cell.html]: Removed

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.js]: Updated

File: client/src/modules/stock/lots-duplicates/duplicate-lots.html
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.html]: Removed

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.js]: Removed.

File: client/src/modules/stock/lots-duplicates/duplicate-lots.js
Comment (jmcameron) [reply client/src/modules/stock/lots-duplicates/duplicate-lots.js]: Updated.

**Pull request merged**
